File: library/src/main/java/com/alamkanak/weekview/drawing/DayLabelDrawer.java
package com.alamkanak.weekview.drawing;

import android.graphics.Canvas;
import android.graphics.Paint;

import com.alamkanak.weekview.model.WeekViewConfig;

import java.util.Calendar;
import java.util.List;

import static com.alamkanak.weekview.utils.DateUtils.isSameDay;
import static com.alamkanak.weekview.utils.DateUtils.today;

public class DayLabelDrawer {

    private WeekViewConfig config;
    private WeekViewDrawingConfig drawingConfig;

    public DayLabelDrawer(WeekViewConfig config) {
        this.config = config;
        this.drawingConfig = config.drawingConfig;
    }

    public void draw(List<Calendar> dayRange, float startPixel, Canvas canvas) {
        for (Calendar day : dayRange) {
            draw(day, startPixel, canvas);

            if (config.isSingleDay()) {
                startPixel = startPixel + config.eventMarginHorizontal;
            }

            startPixel += drawingConfig.widthPerDay + config.columnGap;
        }
    }

    public void draw(Calendar day, float startPixel, Canvas canvas) {
        drawLabel(day, startPixel, canvas);
    }

    private void drawLabel(Calendar day, float startPixel, Canvas canvas) {
        Calendar today = today();
        boolean isSameDay = isSameDay(day, today);

        // Draw the day labels.
        String dayLabel = drawingConfig.dateTimeInterpreter.interpretDate(day);
        if (dayLabel == null) {
            throw new IllegalStateException("A DateTimeInterpreter must not return null date");
        }

        float x = startPixel + drawingConfig.widthPerDay / 2;
        float y = drawingConfig.headerTextHeight + config.headerRowPadding;
        Paint textPaint = isSameDay ? drawingConfig.todayHeaderTextPaint : drawingConfig.headerTextPaint;
        canvas.drawText(dayLabel, x, y, textPaint);
    }

}

File: core/src/main/java/com/alamkanak/weekview/DayLabelDrawer.kt
package com.alamkanak.weekview

import android.graphics.Canvas
import android.util.SparseArray
import java.util.Calendar

internal class DayLabelDrawer<T>(
    private val config: WeekViewConfigWrapper,
    private val cache: WeekViewCache<T>
) : CachingDrawer {

    override fun draw(
        drawingContext: DrawingContext,
        canvas: Canvas
    ) {
        val left = config.timeColumnWidth
        val top = 0f
        val right = canvas.width.toFloat()
        val bottom = config.getTotalHeaderHeight()

        canvas.drawInRect(left, top, right, bottom) {
            drawingContext.dateRangeWithStartPixels.forEach { (date, startPixel) ->
                drawLabel(date, startPixel, this)
            }
        }
    }

    private fun drawLabel(day: Calendar, startPixel: Float, canvas: Canvas) {
        val key = day.toEpochDays()
        val dayLabel = cache.dayLabelCache.get(key) { provideAndCacheDayLabel(key, day) }

        val x = startPixel + config.widthPerDay / 2

        val textPaint = if (day.isToday) {
            config.todayHeaderTextPaint
        } else {
            config.headerTextPaint
        }

        if (config.singleLineHeader) {
            val y = config.headerRowPadding.toFloat() - textPaint.ascent()
            canvas.drawText(dayLabel, x, y, textPaint)
        } else {
            // Draw the multi-line header
            val staticLayout = cache.multiLineDayLabelCache.get(key)
            val y = config.headerRowPadding.toFloat()
            canvas.withTranslation(x, y) {
                staticLayout.draw(this)
            }
        }
    }

    private fun provideAndCacheDayLabel(key: Int, day: Calendar): String {
        return config.dateTimeInterpreter.interpretDate(day).also {
            cache.dayLabelCache.put(key, it)
        }
    }

    override fun clear() {
        cache.dayLabelCache.clear()
        cache.multiLineDayLabelCache.clear()
    }

    private fun Canvas.withTranslation(dx: Float, dy: Float, block: Canvas.() -> Unit) {
        save()
        translate(dx, dy)
        block()
        restore()
    }

    private fun <E> SparseArray<E>.get(key: Int, providerIfEmpty: () -> E): E {
        return get(key) ?: providerIfEmpty.invoke()
    }
}
